File: feeder.rb
# rubocop:disable Style/GlobalVars
# rubocop:disable Metrics/MethodLength, Metrics/AbcSize

require 'redis'
require 'oj'
# require 'pry'

# $filename = '/Users/praphan/Desktop/projects/Redis/100lines.json'
$filename = '/Users/praphan/Desktop/projects/Redis/md-message1-20160308.log'
$redis = Redis.new(path: '/tmp/redis.sock', driver: :hiredis)
$cnt = 0
$chunksize = 700
$buf = []

now = Time.new
puts Time.new.strftime('%H:%M:%S.%3N')

def save(tmp)
  $redis.pipelined do
    id = $cnt < $chunksize ? 1 : $cnt - $chunksize
    tmp.each_with_index do |item, index|
      keys = item.keys
      $redis.hmset("logs:#{id + index}",
                   'id', id + index,
                   'seqNo', item['seqNo'],
                   'mesgType', item['mesgType'],
                   'msg', (keys.length > 2 ? item[keys[2]] : ''))
    end
  end
end

def save_regex(tmp)
  if tmp.empty?
    return
  end
  $redis.pipelined do
    id = $cnt < $chunksize ? 1 : $cnt - $chunksize
    tmp.each_with_index do |item, index|
      $redis.hmset("logs:#{id + index}",
                   'id', id + index,
                   'seqNo', item['seqNo'],
                   'mesgType', item['mesgType'],
                   'msg', item['msg'])
    end
  end
end

def feed_oj
  tmp = []
  File.open($filename).each_line do |line|
    $cnt += 1
    tmp << Oj.load(line)
    if $cnt % $chunksize == 0
      save(tmp)
      tmp = []
    end
  end
  save(tmp)
  $redis.set('logs:size', $cnt)
end

def feed_regex
  tmp = []
  # binding.pry
  File.open($filename).each_line do |line|
    $cnt += 1
    m = line.match(/^{[^{]+({.*})\s*}\s*$/)
    s = if m
          m.captures
        else
          ''
        end
    tmp << {
      'id' => $cnt,
      'seqNo' => line.match(/seqNo[":\s]+(\w+)/i).captures,
      'mesgType' => line.match(/mesgType[":\s]+(\w+)/i).captures,
      'msg' => s
    }

    if $cnt % $chunksize == 0
      save_regex(tmp)
      tmp = []
    end
  end
  # binding.pry
  save_regex(tmp)
  $redis.set('logs:size', $cnt)
end

# feed_oj
feed_regex

puts $cnt
puts Time.new - now
puts Time.new.strftime('%H:%M:%S.%3N')

File: feeder.js
"use strict";

var numeral = require("numeral"),
  moment = require("moment"),
  redis = require("redis"),
  now = moment(),
  filename = "/Users/praphan/Desktop/projects/Redis/md-message1-20160308.log",
  // filename = "/Users/praphan/Desktop/projects/Redis/100lines.json",
  cnt = 0,
  loopsize = 1000,
  data = '';

function save_redis(data) {
  var client = redis.createClient("/tmp/redis.sock"),
    m = client.multi(),
    id = cnt - data.length + 1;
  for (var i = 0; i < data.length; i++) {
    var keys = Object.keys(data[i]),
      s = '';
    if (keys.length > 2) {
      s = JSON.stringify(data[i][keys[2]]);
    }
    m.hmset(
      `logs:${id + i}`,
      "seqNo", data[i].seqNo,
      "mesgType", data[i].mesgType,
      "msg", s
    );
  }
  m.exec();
  client.quit();
}

function process_line_JSONParse(line) {
  if (cnt % loopsize == 1) {
    data = '[' + line;
  } else {
    data = data + ',' + line;
  }
}

// /seqNo[":\s]+(\w+)/i.exec(s)[1]
// /mesgType[":\s]+(\w+)/i.exec(s)[1]
// /}\s*}\s*$/.exec(s)
// /^{[^{]+({.*})\s*}\*$/.exec(s)[1]
var r_seqNo = /seqNo[":\s]+(\w+)/i,
  r_mesgType = /mesgType[":\s]+(\w+)/i,
  r_hasmsg = /}\s*}\s*$/,
  r_msg = /^{[^{]+({.*})\s*}\s*$/;

function process_line_regex(line) {
  if (cnt % loopsize == 1) {
    data = [];
  }
  var a = {
    id: cnt,
    seqNo: r_seqNo.exec(line)[1],
    mesgType: r_mesgType.exec(line)[1]
  };
  if (r_hasmsg.exec(line)) {
    a.msg = r_msg.exec(line)[1];
  } else {
    a.msg = '';
  }
  data.push(a);
}

function save_JSONParse() {
  if (data) {
    save_redis(JSON.parse(data + ']'));
    data = null;
  }
}

function save_regex() {
  if (data) {
    var client = redis.createClient("/tmp/redis.sock"),
      m = client.multi();
    for (var i = 0; i < data.length; i++) {
      m.hmset(
        'logs:' + data[i].id,
        "seqNo", data[i].seqNo,
        "mesgType", data[i].mesgType,
        "msg", data[i].msg
      );
    }
    m.exec();
    client.quit();
    data = null;
  }
}

function load(process_line, save) {
  console.log(moment().format('HH:mm:ss.SSS'));

  require("readline").createInterface({
    input: require("fs").createReadStream(filename)
  }).on("line", function(line) {
    cnt = cnt + 1;
    process_line(line);
    if (cnt % loopsize == 0) {
      // save_ioredis(JSON.parse(data + ']'));
      save();
    }
  }).on("close", function() {
    save();
    var client = redis.createClient("/tmp/redis.sock");
    client.set('logs:size', cnt);
    client.quit();

    console.log("line = " + numeral(cnt).format());
    console.log("chunk = " + loopsize);
    console.log("time = " + moment(moment().diff(now)).format("mm:ss.SSS"));
    console.log(moment().format('HH:mm:ss.SSS'));
  });

}

// load(process_line_JSONParse, save_JSONParse);
load(process_line_regex, save_regex);
